tests/test_create_layout.py

```diff
@@ -0,0 +1,183 @@
+        'one.tgz', 'three.txt']))
+
+
+  def test_new_filesystem_snapshot(self):
+    after = {
+      'five.tgz': '1234567890defghi',
+      'foo/bar/six.tgz': '0000001111112234',
+      'foofoo/seven.txt': '1111222233334555'
+    }
+
+    changes = create_layout.changes_between_snapshots(self.before, after)
+    self.assertEqual(changes, ([], [], ['five.tgz', 'foo/bar/six.tgz',
+        'foofoo/seven.txt'], ['bar/bat/four.tgz', 'foo/two.tgz', 'one.tgz',
+        'three.txt']))
+
+
+  def test_fully_modified_filesystem_snapshot(self):
+
+    after = {
+      'one.tgz': '1234567890aabbcc',
+      'foo/two.tgz': '0000001111112233',
+      'three.txt': '1111222233334455',
+      'bar/bat/four.tgz': '6677889900123456'
+    }
+
+    changes = create_layout.changes_between_snapshots(self.before, after)
+    self.assertEqual(changes, ([], ['bar/bat/four.tgz', 'foo/two.tgz',
+        'one.tgz', 'three.txt'], [], []))
+
+
+  def test_partially_modified_filesystem_snapshot(self):
+
+    after = {
+      'five.txt': '5555555555555555',
+      'one.tgz': '1234567890abcdef',
+      'foo/two.tgz': 'ffffffffffffffff',
+      'bar/bat/four.tgz': '6677889900123456',
+      'baz/six.tgz': '6666666666666666'
+    }
+
+    changes = create_layout.changes_between_snapshots(self.before, after)
+    self.assertEqual(changes, (['one.tgz'], ['bar/bat/four.tgz',
+        'foo/two.tgz'], ['baz/six.tgz', 'five.txt'], ['three.txt']))
+
+
+  def test_create_material_rules_of_initial_step(self):
+    # Zero index means that the current step is the initial step,
+    # so we need to ALLOW all the existing files instead of matching.
+    second_link = in_toto.models.link.Link.read(self.second_step_link_str)
+    links = [second_link]
+    result = True
+
+    expected_materials = [
+      ['ALLOW', 'bar/bat/four.tgz'],
+      ['ALLOW', 'one.tgz'],
+      ['ALLOW', 'foo/two.tgz'],
+      ['ALLOW', 'three.txt'],
+      ['DELETE', 'three.txt'],
+      ['DISALLOW', '*']
+    ]
+
+    for rule in create_layout.create_material_rules(None, second_link):
+      if rule not in expected_materials:
+        result = False
+        break
+
+    self.assertTrue(result)
+
+  def test_create_material_rules_of_not_initial_step(self):
+    # Nonzero index means that the current step is not the initial step,
+    # so we need to MATCH materials with products of the previous step.
+    first_link = in_toto.models.link.Link.read(self.first_step_link_str)
+    second_link = in_toto.models.link.Link.read(self.second_step_link_str)
+    links = [first_link, second_link]
+    result = True
+
+    expected_materials = [
+      ['MATCH', 'bar/bat/four.tgz', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
+      ['MATCH', 'one.tgz', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
+      ['MATCH', 'foo/two.tgz', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
+      ['MATCH', 'three.txt', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
+      ['DELETE', 'three.txt']
```

Comment:
Missing a DISALLOW here, possibly because of the conditional problem I highlighted above. If not, there may be another place to add a DISALLOW! :)